File: apps/elixir/test/getting_started_ml/chapter14_test.exs
defmodule GettingStartedMl.Chapter14Test do
  use ExUnit.Case, async: true

  alias GettingStartedMl.Chapter14, as: C
  alias GettingStartedMl.Dataset
  alias Scholar.Cluster.KMeans

  # 2 次元に離れた 2 つの群れ。
  defp two_groups do
    [[0.0, 0.0], [1.0, 0.0], [0.0, 1.0], [10.0, 10.0], [11.0, 10.0], [10.0, 11.0]]
  end

  # 1 次元に 3 組の点。最適な 3 分割の SSE は 0.5 × 3 = 1.5。
  defp three_pairs, do: [[0.0], [1.0], [10.0], [11.0], [20.0], [21.0]]

  defp spending_table do
    %{
      columns: [:Channel, :Region, :Fresh, :Milk],
      rows: [
        %{Channel: "1", Region: "3", Fresh: "100", Milk: "200"},
        %{Channel: "2", Region: "3", Fresh: "300", Milk: "400"}
      ]
    }
  end

  describe "距離と割り当て" do
    test "2 点間の距離の 2 乗を求める" do
      assert C.squared_distance([0.0, 0.0], [3.0, 4.0]) == 25.0
    end

    test "1 次元の点を最も近い中心に割り当てる" do
      assert C.assign_clusters([[0.0], [1.0], [10.0], [11.0]], [[0.5], [10.5]]) == [0, 0, 1, 1]
    end

    test "2 次元の点をユークリッド距離で割り当てる" do
      assert C.assign_clusters(two_groups(), [[0.0, 0.0], [10.0, 10.0]]) == [0, 0, 0, 1, 1, 1]
    end

    test "距離が同じなら先に並ぶ中心を選ぶ" do
      assert C.assign_clusters([[1.0]], [[0.0], [2.0]]) == [0]
    end
  end

  describe "中心の更新" do
    test "平均が新しい中心になる" do
      assert C.update_centers([[0.0], [1.0], [10.0], [11.0]], [0, 0, 1, 1], [[0.0], [10.0]]) ==
               [[0.5], [10.5]]
    end

    test "点が 1 つも割り当てられなかったクラスタは中心を変えない" do
      assert C.update_centers([[0.0], [1.0]], [0, 0], [[0.0], [99.0]]) == [[0.5], [99.0]]
    end
  end

  describe "SSE" do
    test "各点と中心の距離の 2 乗を合計する" do
      assert C.sum_of_squared_errors([[0.0], [2.0]], [0, 0], [[1.0]]) == 2.0
    end
  end

  describe "繰り返し" do
    test "中心が変わらなくなるまで繰り返す" do
      result = C.fit(two_groups(), [[0.0, 0.0], [10.0, 10.0]])

      assert result.labels == [0, 0, 0, 1, 1, 1]
      assert_in_delta result.sse, 8 / 3, 1.0e-9
    end

    test "最大反復回数に達したら収束していなくても打ち切る" do
      once = C.fit(three_pairs(), [[0.0], [2.0]], 1)
      none = C.fit(three_pairs(), [[0.0], [2.0]], 0)

      assert once.centers == [[0.5], [15.5]]
      assert none.centers == [[0.0], [2.0]]
      assert none.sse > once.sse
    end

    test "初期中心によっては局所解に陥る" do
      assert_in_delta C.fit(three_pairs(), [[0.0], [1.0], [10.0]]).sse, 101.0, 1.0e-9
    end

    test "複数の初期中心の候補のうち SSE が最小の結果を返す" do
      result = C.best(three_pairs(), [[[0.0], [1.0], [10.0]], [[0.0], [10.0], [20.0]]])

      assert_in_delta result.sse, 1.5, 1.0e-9
      assert result.centers == [[0.5], [10.5], [20.5]]
    end
  end

  describe "初期中心" do
    test "シードで選んだ点そのものを初期中心にする" do
      centers = C.choose_initial_centers(three_pairs(), 3, 0)

      assert length(centers) == 3
      assert Enum.all?(centers, &(&1 in three_pairs()))
      assert Enum.uniq(centers) == centers
    end

    test "同じシードなら同じ初期中心を選ぶ" do
      assert C.choose_initial_centers(three_pairs(), 3, 7) ==
               C.choose_initial_centers(three_pairs(), 3, 7)
    end
  end

  describe "エルボー法" do
    test "クラスタ数ごとの SSE をクラスタ数の順に返す" do
      result = C.sse_by_cluster_count(three_pairs(), 1..3, 0)

      assert Enum.map(result, &elem(&1, 0)) == [1, 2, 3]
      assert_in_delta elem(Enum.at(result, 2), 1), 1.5, 1.0e-9
    end

    test "クラスタ数を増やすほど SSE が小さくなる" do
      sse = Enum.map(C.sse_by_cluster_count(three_pairs(), 1..4, 0), &elem(&1, 1))

      assert Enum.zip(sse, tl(sse)) |> Enum.all?(fn {a, b} -> b < a end)
    end
  end

  describe "Scholar の KMeans" do
    test "初期中心そのものは渡せない" do
      # :init に渡せるのは :k_means_plus_plus と :random の 2 つだけ
      assert_raise NimbleOptions.ValidationError, fn ->
        KMeans.fit(Nx.tensor(two_groups(), type: :f64),
          num_clusters: 2,
          init: [[0.0, 0.0], [10.0, 10.0]]
        )
      end
    end

    test "離れた 2 つの群れなら自作と同じ中心にたどり着く" do
      theirs = Enum.sort(C.scholar_centers(two_groups(), 2, 0))
      mine = Enum.sort(C.fit_with_restarts(two_groups(), 2, 0).centers)

      assert Enum.zip(mine, theirs)
             |> Enum.all?(fn {a, b} -> C.squared_distance(a, b) < 1.0e-9 end)
    end

    test "同じ定義の SSE で比べる" do
      assert_in_delta C.scholar_sse(two_groups(), 2, 0), 8 / 3, 1.0e-9
    end
  end

  describe "支出額の読み込み" do
    test "Channel と Region を除いた列を読み込む" do
      %{columns: columns, x: x} = C.spending(spending_table())

      assert columns == [:Fresh, :Milk]
      assert x == [%{Fresh: 100.0, Milk: 200.0}, %{Fresh: 300.0, Milk: 400.0}]
    end

    test "空欄があれば失敗する" do
      table = %{columns: [:Fresh], rows: [%{Fresh: ""}]}

      assert_raise ArgumentError, fn -> C.spending(table) end
    end

    test "標準化すると列ごとの平均が 0・標準偏差が 1 になる" do
      %{columns: columns, x: x} = C.spending(spending_table())
      points = C.standardize(x, columns)

      assert points == [[-1.0, -1.0], [1.0, 1.0]]
    end
  end

  describe "クラスタごとの要約" do
    test "件数の多い順に、元の単位での平均を並べる" do
      x = [%{Fresh: 10.0}, %{Fresh: 20.0}, %{Fresh: 100.0}]

      assert C.summarize_clusters(x, [:Fresh], [1, 1, 0]) == [
               %{cluster: 1, count: 2, means: %{Fresh: 15.0}},
               %{cluster: 0, count: 1, means: %{Fresh: 100.0}}
             ]
    end

    test "件数が同じならクラスタ番号の昇順に並べる" do
      x = [%{Fresh: 10.0}, %{Fresh: 20.0}]

      assert Enum.map(C.summarize_clusters(x, [:Fresh], [1, 0]), & &1.cluster) == [0, 1]
    end
  end

  describe "実データ" do
    @tag :data
    test "440 件 6 列を読み込む" do
      %{columns: columns, x: x} = wholesale()

      assert length(x) == 440
      assert columns == [:Fresh, :Milk, :Grocery, :Frozen, :Detergents_Paper, :Delicassen]
    end

    @tag :data
    test "標準化したデータのクラスタ数 1 の SSE は件数と列数の積になる" do
      # 件数で割る標準偏差で標準化したので、列ごとの分散は 1 になる
      %{columns: columns, x: x} = wholesale()
      [{1, sse}] = C.sse_by_cluster_count(C.standardize(x, columns), [1], 0)

      assert_in_delta sse, 440.0 * 6, 1.0e-6
    end

    @tag :data
    test "クラスタ数を増やすほど SSE が小さくなる" do
      %{columns: columns, x: x} = wholesale()
      sse = Enum.map(C.sse_by_cluster_count(C.standardize(x, columns), 1..10, 0), &elem(&1, 1))

      assert Enum.zip(sse, tl(sse)) |> Enum.all?(fn {a, b} -> b < a end)
    end

    # Scholar の KMeans を 10 通りのクラスタ数で学習するので、既定の 60 秒では足りない
    @tag :data
    @tag timeout: 600_000
    test "実行するとエルボー法の SSE とクラスタごとの特徴を表示する" do
      output = ExUnit.CaptureIO.capture_io(&C.run/0)

      assert String.split(output, "\n", trim: true) == [
               "データ件数: 440（支出額 6 列）",
               "クラスタ数ごとの SSE（初期中心 10 通りの最小値）:",
               "クラスタ数\t自作\tScholar（k-means++）",
               "1\t2640.00\t2640.00",
               "2\t1954.18\t1954.73",
               "3\t1614.52\t1632.87",
               "4\t1334.36\t1329.41",
               "5\t1085.27\t1075.51",
               "6\t947.20\t923.05",
               "7\t888.22\t829.72",
               "8\t775.24\t753.01",
               "9\t690.81\t676.57",
               "10\t618.17\t602.40",
               "クラスタ数 5 のクラスタごとの件数と平均支出額:",
               "クラスタ\t件数\tFresh\tMilk\tGrocery\tFrozen\tDetergents_Paper\tDelicassen",
               "2\t265\t8909\t2967\t3804\t2248\t989\t962",
               "1\t96\t5509\t10556\t16478\t1420\t7199\t1659",
               "0\t65\t31117\t4260\t5374\t7225\t849\t2286",
               "3\t10\t15965\t34709\t48537\t3055\t24875\t2943",
               "4\t4\t52022\t31696\t18491\t29826\t2699\t19656"
             ]
    end
  end

  defp wholesale, do: C.load_spending(Path.join(Dataset.dir(), "Wholesale.csv"))
end

File: apps/clojure/test/getting_started_ml/chapter14_test.clj
(ns getting-started-ml.chapter14-test
  (:require [clojure.test :refer [deftest is testing]]
            [getting-started-ml.chapter14 :as ch])
  (:import [org.tribuo.clustering.kmeans KMeansTrainer KMeansTrainer$Initialisation]))

(defn- close-to?
  "小数の誤差を許して比べる。"
  ([a b] (close-to? a b 1e-9))
  ([a b tolerance] (< (abs (- a b)) tolerance)))

(defn- two-groups
  "0 の近くと 10 の近くに、はっきり分かれた 2 グループ。"
  []
  [[0.0 0.0] [0.0 1.0] [1.0 0.0] [10.0 10.0] [10.0 11.0] [11.0 10.0]])

(defn- two-pairs
  "1 次元に 2 組の点を並べた、クラスタ数がはっきり決まるデータ。"
  []
  [[0.0] [2.0] [10.0] [12.0]])

(defn- three-pairs
  "1 次元に 3 組の点を並べたデータ。"
  []
  [[0.0] [1.0] [10.0] [11.0] [20.0] [21.0]])

;; ## 割り当てと中心の更新

(deftest 各点を最も近い中心のクラスタに割り当てる
  (testing "1 次元の点"
    (is (= [0 0 1 1] (ch/assign-clusters [[0.0] [1.0] [10.0] [11.0]] [[0.5] [10.5]]))))
  (testing "2 次元の点をユークリッド距離で割り当てる"
    (is (= [0 0 0 1 1 1] (ch/assign-clusters (two-groups) [[0.0 0.0] [10.0 10.0]]))))
  (testing "距離が同じなら先に並ぶ中心を選ぶ"
    (is (= [0] (ch/assign-clusters [[1.0]] [[0.0] [2.0]])))))

(deftest 距離の二乗を求める
  (is (close-to? 25.0 (ch/squared-distance [0.0 0.0] [3.0 4.0]))))

(deftest クラスタごとに割り当てられた点の平均を新しい中心にする
  (testing "平均が新しい中心になる"
    (is (= [[0.5] [10.5]]
           (ch/update-centers [[0.0] [1.0] [10.0] [11.0]] [0 0 1 1] [[0.0] [10.0]]))))
  (testing "点が 1 つも割り当てられなかったクラスタは中心を変えない"
    (is (= [[0.5] [99.0]]
           (ch/update-centers [[0.0] [1.0]] [0 0] [[0.0] [99.0]])))))

;; ## 誤差平方和

(deftest 各点と所属するクラスタの中心との距離の二乗を合計する
  (testing "中心にぴったり重なれば 0"
    (is (close-to? 0.0 (ch/sum-of-squared-errors [[1.0] [2.0]] [0 1] [[1.0] [2.0]]))))
  (testing "中心から離れた点ほど誤差が大きくなる"
    (is (close-to? 0.5 (ch/sum-of-squared-errors [[0.0] [1.0]] [0 0] [[0.5]])))
    (is (close-to? 4.0 (ch/sum-of-squared-errors [[-1.0] [1.0]] [0 0] [[1.0]])))))

;; ## 収束まで繰り返す

(deftest 割り当てが変わらなくなるまで割り当てと中心の更新を繰り返す
  (let [result (ch/fit (two-groups) [[0.0 0.0] [1.0 1.0]])]
    (is (= [0 0 0 1 1 1] (:labels result)))
    (is (= [[(/ 1.0 3) (/ 1.0 3)] [(/ 31.0 3) (/ 31.0 3)]]
           (mapv #(mapv double %) (:centers result))))))

(deftest 最大反復回数に達したら収束していなくても打ち切る
  (let [once (ch/fit (three-pairs) [[0.0] [2.0]] 1)
        none (ch/fit (three-pairs) [[0.0] [2.0]] 0)]
    (is (= [[0.5] [15.5]] (mapv #(mapv double %) (:centers once))))
    (is (= [[0.0] [2.0]] (:centers none)))
    (is (> (:sse none) (:sse once)))))

;; ## 初期中心

(deftest データの中から重複なくクラスタ数だけ点を選ぶ
  (let [centers (ch/choose-initial-centers (three-pairs) 3 0)]
    (is (= 3 (count centers)))
    (is (= 3 (count (distinct centers))))
    (is (every? (set (three-pairs)) centers))))

(deftest 同じシードなら同じ点を選びシードが違えば違う点を選ぶ
  (is (= (ch/choose-initial-centers (three-pairs) 3 0)
         (ch/choose-initial-centers (three-pairs) 3 0)))
  (is (not= (ch/choose-initial-centers (three-pairs) 3 0)
            (ch/choose-initial-centers (three-pairs) 3 7))))

;; ## エルボー法と局所解

(deftest クラスタ数ごとにクラスタリングしたときのSSEを求める
  (is (= [[1 104.0] [2 4.0]] (ch/sse-by-cluster-count (two-pairs) [1 2] 0))))

(deftest 初期中心によっては局所解に陥る
  (is (close-to? 101.0 (:sse (ch/fit (three-pairs) [[0.0] [1.0] [10.0]])))))

(deftest 複数の初期中心の候補のうちSSEが最小の結果を返す
  (let [result (ch/best (three-pairs) [[[0.0] [1.0] [10.0]] [[0.0] [10.0] [20.0]]])]
    (is (close-to? 1.5 (:sse result)))
    (is (= [[0.5] [10.5] [20.5]] (mapv #(mapv double %) (:centers result))))))

(deftest 初期中心を変えて繰り返し最小のSSEを使う
  (is (close-to? 1.5 (:sse (ch/fit-with-restarts (three-pairs) 3 0 10)))))

;; ## Tribuo

(deftest KMeansTrainerの初期化方法はRANDOMとPLUSPLUSだけ
  (is (= ["RANDOM" "PLUSPLUS"] (mapv #(.name %) (KMeansTrainer$Initialisation/values))))
  (is (not-any? (fn [constructor]
                  (some #(or (.isArray ^Class %)
                             (.isAssignableFrom java.util.Collection ^Class %))
                        (.getParameterTypes constructor)))
                (.getConstructors KMeansTrainer))))

(deftest はっきり分かれた二グループならTribuoのSSEも自作と同じになる
  (is (close-to? (:sse (ch/fit-with-restarts (two-groups) 2 0 3))
                 (ch/tribuo-best-sse (two-groups) 2 0 3))))

;; ## 支出額

(defn- spending-table
  "Wholesale.csv に似た小さな表。"
  []
  {:columns [:Channel :Region :Fresh :Milk]
   :rows [{:Channel "1" :Region "3" :Fresh "100" :Milk "200"}
          {:Channel "2" :Region "3" :Fresh "300" :Milk "400"}
          {:Channel "1" :Region "1" :Fresh "500" :Milk "600"}]})

(deftest ChannelとRegionを除いた支出額の列を読み込む
  (let [{:keys [columns x]} (ch/spending (spending-table))]
    (is (= [:Fresh :Milk] columns))
    (is (= [100.0 300.0 500.0] (mapv :Fresh x)))))

(deftest 列ごとに平均零と標準偏差一の点に変換する
  (let [{:keys [columns x]} (ch/spending (spending-table))
        points (ch/standardize x columns)]
    (is (= 3 (count points)))
    (doseq [j (range 2)]
      (let [values (mapv #(nth % j) points)
            mean (/ (reduce + values) (count values))]
        (is (close-to? 0.0 mean))
        (is (close-to? 1.0 (Math/sqrt (/ (reduce + (map #(* % %) values)) (count values)))))))))

(deftest クラスタごとの件数と平均を件数の多い順に並べる
  (let [{:keys [columns x]} (ch/spending (spending-table))
        summaries (ch/summarize-clusters x columns [1 0 1])]
    (is (= [1 0] (mapv :cluster summaries)))
    (is (= [2 1] (mapv :count summaries)))
    (is (close-to? 300.0 (get-in (first summaries) [:means :Fresh])))))
